### ISSUE TITLE
Missing role relations on user

### ISSUE THREAD

**Issue by elbird** [gentics/mesh#255]
## Gentics Mesh Version, operating system, or hardware.
- v0.13.3

## Problem
A mesh user could not access some nodes because she was missing a role relation. The user
("uuid": "e05f3a2def1e4bc29f3a2def1edbc21d") is assign to the group ("Role 1",	 "uuid": "4121eac4f84a4fc4a1eac4f84aefc48b") which in turn has 2 roles ("Common" "uuid": "f6ed1d618d284929ad1d618d2899297b" and "Role1" "uuid": "750b8b805aa84b108b8b805aa86b1077"). When the user is logged in Mesh it appears as if one role ("Common") is missing and resources where the permissions are set through this role are not accessible. Analysis of the raw database found that the user was missing the automatically generated direct relation from user to role.

Example Permission check

The user has no permission on a node (uuid: 211fe68034d044899fe68034d0f489e4)
/users/e05f3a2def1e4bc29f3a2def1edbc21d/permissions/projects/2358da6c73704d8f98da6c73701d8f2d/nodes/211fe68034d044899fe68034d0f489e4

````
{
	"create": false,
	"read": false,
	"update": false,
	"delete": false,
	"publish": false,
	"readPublished": false
}
````
But on the role the "read" and "readPublished" permissions are set:
/roles/f6ed1d618d284929ad1d618d2899297b/permissions/projects/2358da6c73704d8f98da6c73701d8f2d/nodes/211fe68034d044899fe68034d0f489e4
````

{
	"create": false,
	"read": true,
	"update": false,
	"delete": false,
	"publish": false,
	"readPublished": true
}
````

**Comment by Jotschi** [comment]
* User: id:193:1, uuid: e05f3a
* Group: uuid: 4121ea
* RoleA: id: 191:1, uuid: f6ed1
* RoleB: id: 192:1, uuid: 750b8b
* Node: id: 198:1, uuid: 211fe68

Edge between user and Role 191:1 via "ASSIGNED_TO_ROLE" edge was missing.

**Comment by Jotschi** [comment]
TODO: 
* [ ] Check whether delayed events could cause such issue
* [ ] Add consistency check to detect such problems
* [ ] Check implementation for possible bugs
* [ ] Add tests which will work in combination with the consistency check

**Comment by padabou** [comment]
I had an issue to with user/role/group since last release.
When i affect a group to a user, i get the following error : 
`mesh-node1_1  | 16:47:32.869 [node1] ERROR [vert.x-eventloop-thread-6] - Error for request in path: /mesh-ui/?null
mesh-node1_1  | 16:47:32.874 [node1] ERROR [vert.x-eventloop-thread-6] - Error:
mesh-node1_1  | java.nio.channels.ClosedChannelException: null
mesh-node1_1  |         at io.netty.channel.AbstractChannel$AbstractUnsafe.write(...)(Unknown Source) ~[mesh.jar:na]
mesh-node1_1  | 16:47:32.874 [node1] ERROR [vert.x-eventloop-thread-0] - Error:
mesh-node1_1  | java.nio.channels.ClosedChannelException: null
mesh-node1_1  |         at io.netty.channel.AbstractChannel$AbstractUnsafe.write(...)(Unknown Source) ~[mesh.jar:na]
mesh-node1_1  | 16:47:32.877 [node1] ERROR [vert.x-eventloop-thread-0] - Unexpected exception in route
mesh-node1_1  | java.lang.IllegalStateException: Response has already been written
mesh-node1_1  |         at io.vertx.core.http.impl.HttpServerResponseImpl.checkWritten(HttpServerResponseImpl.java:572) ~[mesh.jar:na]
mesh-node1_1  |         at io.vertx.core.http.impl.HttpServerResponseImpl.putHeader(HttpServerResponseImpl.java:159) ~[mesh.jar:na]
mesh-node1_1  |         at io.vertx.core.http.impl.HttpServerResponseImpl.putHeader(HttpServerResponseImpl.java:58) ~[mesh.jar:na]
mesh-node1_1  |         at com.gentics.mesh.router.route.FailureHandler.handle(FailureHandler.java:105) ~[mesh.jar:na]
mesh-node1_1  |         at com.gentics.mesh.router.route.FailureHandler.handle(FailureHandler.java:26) ~[mesh.jar:na]
mesh-node1_1  |         at io.vertx.ext.web.impl.RouteImpl.handleFailure(RouteImpl.java:230) ~[mesh.jar:na]
mesh-node1_1  |         at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:99) ~[mesh.jar:na]
mesh-node1_1  |         at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:120) [mesh.jar:na]
mesh-node1_1  |         at io.vertx.ext.web.impl.RoutingContextImpl.doFail(RoutingContextImpl.java:452) [mesh.jar:na]
mesh-node1_1  |         at io.vertx.ext.web.impl.RoutingContextImpl.fail(RoutingContextImpl.java:153) [mesh.jar:na]
mesh-node1_1  |         at io.vertx.ext.web.impl.RoutingContextWrapper.fail(RoutingContextWrapper.java:73) [mesh.jar:na]
mesh-node1_1  |         at io.vertx.ext.web.handler.impl.StaticHandlerImpl.lambda$null$7(StaticHandlerImpl.java:409) [mesh.jar:na]
mesh-node1_1  |         at io.vertx.core.http.impl.HttpServerResponseImpl.lambda$null$3(HttpServerResponseImpl.java:519) ~[mesh.jar:na]
mesh-node1_1  |         at io.vertx.core.impl.ContextImpl.lambda$wrapTask$2(ContextImpl.java:344) ~[mesh.jar:na]
mesh-node1_1  |         at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163) ~[mesh.jar:na]
mesh-node1_1  |         at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:403) ~[mesh.jar:na]
mesh-node1_1  |         at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:463) ~[mesh.jar:na]
mesh-node1_1  |         at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858) ~[mesh.jar:na]
mesh-node1_1  |         at java.lang.Thread.run(Thread.java:745) ~[na:1.8.0_111-internal]
mesh-node1_1  | 16:47:32.878 [node1] ERROR [vert.x-eventloop-thread-6] - Unexpected exception in route
mesh-node1_1  | java.lang.IllegalStateException: Response has already been written
mesh-node1_1  |         at io.vertx.core.http.impl.HttpServerResponseImpl.checkWritten(HttpServerResponseImpl.java:572) ~[mesh.jar:na]
mesh-node1_1  |         at io.vertx.core.http.impl.HttpServerResponseImpl.putHeader(HttpServerResponseImpl.java:159) ~[mesh.jar:na]
mesh-node1_1  |         at io.vertx.core.http.impl.HttpServerResponseImpl.putHeader(HttpServerResponseImpl.java:58) ~[mesh.jar:na]
mesh-node1_1  |         at com.gentics.mesh.router.route.FailureHandler.handle(FailureHandler.java:105) ~[mesh.jar:na]
mesh-node1_1  |         at com.gentics.mesh.router.route.FailureHandler.handle(FailureHandler.java:26) ~[mesh.jar:na]
mesh-node1_1  |         at io.vertx.ext.web.impl.RouteImpl.handleFailure(RouteImpl.java:230) ~[mesh.jar:na]
mesh-node1_1  |         at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:99) ~[mesh.jar:na]
mesh-node1_1  |         at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:120) [mesh.jar:na]
mesh-node1_1  |         at io.vertx.ext.web.impl.RoutingContextImpl.doFail(RoutingContextImpl.java:452) [mesh.jar:na]
mesh-node1_1  |         at io.vertx.ext.web.impl.RoutingContextImpl.fail(RoutingContextImpl.java:153) [mesh.jar:na]
mesh-node1_1  |         at io.vertx.ext.web.impl.RoutingContextWrapper.fail(RoutingContextWrapper.java:73) [mesh.jar:na]
mesh-node1_1  |         at io.vertx.ext.web.handler.impl.StaticHandlerImpl.lambda$null$7(StaticHandlerImpl.java:409) [mesh.jar:na]
mesh-node1_1  |         at io.vertx.core.http.impl.HttpServerResponseImpl.lambda$null$3(HttpServerResponseImpl.java:519) ~[mesh.jar:na]
mesh-node1_1  |         at io.vertx.core.impl.ContextImpl.lambda$wrapTask$2(ContextImpl.java:344) ~[mesh.jar:na]
mesh-node1_1  |         at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163) ~[mesh.jar:na]
mesh-node1_1  |         at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:403) ~[mesh.jar:na]
mesh-node1_1  |         at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:463) ~[mesh.jar:na]
mesh-node1_1  |         at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858) ~[mesh.jar:na]
mesh-node1_1  |         at java.lang.Thread.run(Thread.java:745) ~[na:1.8.0_111-internal]
mesh-node1_1  | 16:47:32.889 [node1] INFO  [vert.x-eventloop-thread-2] - 172.19.0.1 - - [Mon, 15 Jan 2018 16:47:32 GMT] "GET /mesh-ui/ HTTP/1.1" 200 657 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/63.0.3239.132 Safari/537.36"
mesh-node1_1  | 16:47:32.907 [node1] INFO  [vert.x-eventloop-thread-2] - 172.19.0.1 - - [Mon, 15 Jan 2018 16:47:32 GMT] "GET /mesh-ui/mesh-ui-config.js HTTP/1.1" 200 2905 "http://localhost:8080/mesh-ui/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/63.0.3239.132 Safari/537.36"
mesh-node1_1  | 16:47:33.286 [node1] ERROR [vert.x-worker-thread-12] - Error handling transaction
mesh-node1_1  | com.gentics.mesh.core.rest.error.NotModifiedException: null
`
But my browser display a message to say everything right ... 

Maybe there is relation between our issue.

**Comment by Jotschi** [comment]
@padabou Sometimes the UI does not forward errors correctly. Maybe the request failed. You could perhaps check the browser network tab? I'll check whether the issues could be related. I'll keep you posted.

**Comment by padabou** [comment]
@Jotschi, i follow your advice, but network tab said everything right ... 
So i try a new test : 
Start an empty cms
Create a user 
Add a group to user

It doesn't work at the third step : browser said everything right but it's not.
When i refresh my page group wasn't affected to user.

I attach the log from cms starting (inside docker container) to page refreshing.

[getmesh.log](https://github.com/gentics/mesh/files/1634274/getmesh.log)

**Comment by Jotschi** [comment]
@padabou Thanks for the log. I don't see anything wrong at the server side. Could you perhaps create a zip of the data/graphdb folder and upload it somewhere? I could then check whether that dump contains the same inconsistency that we encountered before.

**Comment by padabou** [comment]
I push my directories data / config on https://we.tl/6OVqoYtPLw

**Comment by Jotschi** [comment]
@padabou Just checked your data. There is no user/role inconsistency within your graph data. But I'll compare the log against the actual graph data and see whether I can spot any problems.

**Comment by Jotschi** [comment]
Small update: The added consistency checks are being executed after each test to assert the database structure. An inconsistency was detected when deleting a group. I'll fix that bug within the next update.

**Comment by Jotschi** [comment]
Fix has been released in 0.14.1 - Closing this issue
